### ISSUE TITLE
Add check for all nan slices to preprocessing

### ISSUE THREAD

**Issue by jbusecke** [m2lines/ocean_emulators#34]
I think there should be no case where all values are nan across the 'x' and 'y' dimensions. 

We should consider adding a check to the pre/postprocessing `_data_test` functions that checks for this.

We can test this by calculating:
```python
all_nan_test = np.isnan(ds_input.to_array()).all(['x','y']).load()
```
But this will chew through the entire dataset!

This would detect if all values are nan, there might also be more tricky scenarios where the nans across time/variable do not match, maybe that is a better test?

**Comment by jbusecke** [comment]
Here is a functioning snippet to add to #12:
```python
all_nan_test = np.isnan(ds_input.to_array()).all(['x','y']).load()
all_dims = ['lev', 'time', 'variable']
nans_found_index = {}
for dim in all_dims:
    other_dims = [di for di in all_dims if di != dim]
    test = all_nan_test.any(other_dims)
    index = all_nan_test[dim].isel({dim:test})
    nans_found_index[dim] = index.data

print(f"Found all nan slices in a subset of the data: {nans_found_index}")
```

**Comment by jbusecke** [comment]
More exhaustive test snippet that tests 2d, 3d variables separately, then cross checks them against each other. For any failure it gives the indexers where in the dataset to find the non matching nan-values: 
```python
# i need to test 2d and 3d separately
def split_2d_3d(ds:xr.Dataset):
    ds_2d = xr.Dataset({v:ds[v] for v in ds.data_vars if 'lev' not in ds[v].dims})
    ds_3d = xr.Dataset({v:ds[v] for v in ds.data_vars if 'lev' in ds[v].dims})
    return ds_2d, ds_3d

def find_index_for_true(da_bool: xr.DataArray):
    """Find slices along all dimensions within a boolean array that have any True value"""
    all_dims = da_bool.dims
    true_found_index = {}
    for dim in all_dims:
        other_dims = [di for di in all_dims if di != dim]
        test = da_bool.any(other_dims).load()
        index = da_bool[dim].isel({dim:test})
        true_found_index[dim] = index.data
    return true_found_index 

def test_nan_consistency(ds:xr.Dataset, name='None'):
    """Test the consistency of nan values in the dataset across variables and time 
    (compared to a reference at time=0)."""
    ds = ds.to_array()
    ref = ds.isel(time=0)
    # # make sure the ref data has nans in the same places for all variables
    a = (np.isnan(ref.isel(variable=0)) != np.isnan(ref)).all(['variable'])
    # find the index values for true values in b
    index = find_index_for_true(a)
    if not all(len(v) == 0 for v in index.values()):
        raise ValueError(f"Found non-matching nan values between variables on the first time step.")
    
    ## make sure that the ref nan pattern is the same as every time step
    b = np.isnan(ref) != np.isnan(ds)
    
    # find the index values for true values in b
    index = find_index_for_true(b)
    
    # if they are all length 0 all is good, otherwise raise.
    if not all(len(v) == 0 for v in index.values()):
        raise ValueError(f"{name}:Found nonmatching nans compared to first time step in the following indexes {index}")

ds_nan_test_2d, ds_nan_test_3d = split_2d_3d(ds_input) 

test_nan_consistency(ds_nan_test_2d, '2D nan consistency check')
test_nan_consistency(ds_nan_test_3d, '3D nan consistency check')

# if this passes, the last thing to do is to compare the nan concistency between 2d and 3d
# merge the first two timesteps and test again
cross_nan_test = xr.merge([ds_nan_test_2d.isel(time=slice(0,2)), ds_nan_test_3d.isel(lev=0, time=slice(0,2))])
test_nan_consistency(cross_nan_test, 'Cross nan consistency check')
```

**Comment by jbusecke** [comment]
Implemented in #12.
